Handle multiple values in X-Forwarded-Proto

elonazoulay opened the pull request:
Return the first value in X-Forwarded-Proto.
This is because X-Forwarded-For will return the first address.

Fixes #1699

elonazoulay commented:
The following stack trace is from a request with multiple values for X-Forwarded-Proto:
```java.lang.IllegalArgumentException: The string "https,http" for the URI component SCHEME contains an invalid character, ',', at index 5.
        at org.glassfish.jersey.uri.UriComponent.validate(UriComponent.java:163)
        at org.glassfish.jersey.uri.internal.JerseyUriBuilder.scheme(JerseyUriBuilder.java:274)
        at org.glassfish.jersey.uri.internal.JerseyUriBuilder.scheme(JerseyUriBuilder.java:72)
        at io.prestosql.dispatcher.QueuedStatementResource.getQueuedUri(QueuedStatementResource.java:235)
        at io.prestosql.dispatcher.QueuedStatementResource.access$400(QueuedStatementResource.java:86)
        at io.prestosql.dispatcher.QueuedStatementResource$Query.lambda$getNextUri$2(QueuedStatementResource.java:406)
        at java.util.Optional.orElseGet(Optional.java:267)
        at io.prestosql.dispatcher.QueuedStatementResource$Query.getNextUri(QueuedStatementResource.java:406)
        at io.prestosql.dispatcher.QueuedStatementResource$Query.createQueryResults(QueuedStatementResource.java:382)
        at io.prestosql.dispatcher.QueuedStatementResource$Query.getQueryResults(QueuedStatementResource.java:356)
        at io.prestosql.dispatcher.QueuedStatementResource.postStatement(QueuedStatementResource.java:163)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:76)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:148)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:191)
        at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:200)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:103)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:493)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:415)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:104)
        at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:277)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:272)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:268)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:316)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:298)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:268)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:289)
        at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:256)
        at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:703)
        at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:416)
        at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:370)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:389)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:342)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:229)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:867)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1623)
        at io.prestosql.server.security.AuthenticationFilter.doFilter(AuthenticationFilter.java:75)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
        at io.airlift.http.server.TraceTokenFilter.doFilter(TraceTokenFilter.java:63)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
        at io.airlift.http.server.TimingFilter.doFilter(TimingFilter.java:51)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:540)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
        at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:703)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1345)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:480)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1247)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:126)
        at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:174)
        at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:61)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.eclipse.jetty.server.Server.handle(Server.java:502)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:364)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
        at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683)
        at java.lang.Thread.run(Thread.java:748)
```

elonazoulay commented:
Resolves #1699

findepi commented:
thanks @elonazoulay for your PR!

The commit changes two things:

1. makes X-Forwarded-Proto treatment conditional, based on forwardedHeaderSupport
2. supports list in X-Forwarded-Proto

Ad 1:
I didn't do this when adding `forwardedHeaderSupport`. Reason was: there is no security implication if we accept X-Forwarded-Proto send by the client. We don't need to this conditionally. (There are certain security implications of accepting `Forwarded` unconditionally, hence the toggle).
I would prefer not to have this change. Please convince me we need it.

Ad 2:
did you consider leveraging information from `Forwarded` header instead of parsing `X-Forwarded-Proto`?
Is there documentation for `X-Forwarded-Proto` structure?

elonazoulay commented:
Thanks @findepi , makes sense. I removed the conditional behavior and just added support for a list in X-Forwarded-Proto. The gateway we use (modified lyft gateway) appends X-Forwarded-Proto and so does istio which we use as well. I can add support for Forwarded in a separate pull request as this is a standard

findepi commented on presto-main/src/test/java/io/prestosql/dispatcher/TestXForwardedProtoScheme.java:
```diff
@@ -0,0 +1,35 @@
+import static io.prestosql.dispatcher.QueuedStatementResource.getScheme;
+import static org.testng.Assert.assertEquals;
+
+public class TestXForwardedProtoScheme
```
```suggestion
public class TestQueuedStatementResource
```

findepi commented on presto-main/src/test/java/io/prestosql/dispatcher/TestXForwardedProtoScheme.java:
```diff
@@ -0,0 +1,35 @@
+    {
+        UriInfo uriInfo = MockUriInfo.from("http://example.com");
+        assertEquals(getScheme("https", uriInfo), "https");
+        assertEquals(getScheme("https,http", uriInfo), "https");
```
Add a test with
```
getScheme(",https", uriInfo)
```

Now it returns `https`, but maybe we should consider such header as malformed and ignore?
@electrum WDYT?

elonazoulay commented:
@findepi , thanks for the quick review!

findepi commented on presto-main/src/test/java/io/prestosql/dispatcher/TestQueuedStatementResource.java:
```diff
@@ -0,0 +1,36 @@
+        assertEquals(getScheme("https,http", uriInfo), "https");
+        assertEquals(getScheme("", uriInfo), "http");
+        assertEquals(getScheme(null, uriInfo), "http");
+        assertEquals(getScheme(",https", uriInfo), "https");
```
@electrum what do you think about ignoring the header (as malformed) in this case?

elonazoulay commented:
Updated with an additional commit to ensure that getScheme() is used to handle x-forwarded-proto.
This fixed an issue in our deployment where we use a (heavily) modified version of lyft gateway and istio sidecars. Clients connect using "https://presto-gateway" and the gateway proxies the request to "http://presto-server", it adds an extra x-forwarded-proto header which is incorrectly handled as "https,http". 

This will fix the issue for any deployment which sits behind a proxy which adds multiple x-forwarded-proto headers.

findepi commented:
it seems this PR fall thru the cracks, waiting e.g. for some more clarification (eg my last comment)
Elon, could you please rebase?

elonazoulay commented:
Hey, thanks for following up! Should this pull request be closed? 
It looks like 2b39b6f988480e9cc2fc6e8267f67d9ffb03f220 `Replace x-forwarded support with Airlift version` takes care of this. Actually in our production environment we seem to be ok without this since upgrading to 334.

hashhar commented:
@elonazoulay Can you confirm if this is indeed obsolete and no longer required?

elonazoulay commented:
Yep, this is obsolete and can be closed, thanks!
